zrml/hybrid-router/src/benchmarking.rs

```diff
@@ -0,0 +1,249 @@
+    T: Config,
+{
+    let market = Market {
+        base_asset,
+        creation: MarketCreation::Permissionless,
+        creator_fee: Perbill::zero(),
+        creator: caller.clone(),
+        oracle: caller,
+        metadata: vec![0, 50],
+        market_type: MarketType::Categorical(asset_count),
+        period: MarketPeriod::Block(0u32.into()..1u32.into()),
+        deadlines: Default::default(),
+        scoring_rule: ScoringRule::AmmCdaHybrid,
+        status: MarketStatus::Active,
+        report: None,
+        resolved_outcome: None,
+        dispute_mechanism: None,
+        bonds: Default::default(),
+        early_close: None,
+    };
+    let maybe_market_id = T::MarketCommons::push_market(market);
+    maybe_market_id.unwrap()
+}
+
+fn create_market_and_deploy_pool<T: Config>(
+    caller: T::AccountId,
+    base_asset: AssetOf<T>,
+    asset_count: u16,
+    amount: BalanceOf<T>,
+) -> MarketIdOf<T>
+where
+    T: Config,
+{
+    let market_id = create_market::<T>(caller.clone(), base_asset, asset_count);
+    let total_cost = amount + T::AssetManager::minimum_balance(base_asset);
+    assert_ok!(T::AssetManager::deposit(base_asset, &caller, total_cost));
+    assert_ok_with_transaction!(T::CompleteSetOperations::buy_complete_set(
+        caller.clone(),
+        market_id,
+        amount
+    ));
+    assert_ok_with_transaction!(T::AmmPoolDeployer::deploy_pool(
+        caller,
+        market_id,
+        amount,
+        create_spot_prices::<T>(asset_count),
+        CENT.saturated_into(),
+    ));
+    market_id
+}
+
+#[benchmarks]
+mod benchmarks {
+    use super::*;
+
+    #[benchmark]
+    fn buy(n: Linear<2, 128>, o: Linear<0, 100>) {
+        let buyer: T::AccountId = whitelisted_caller();
+        let base_asset = Asset::Ztg;
+        let asset_count = n.try_into().unwrap();
+        let market_id = create_market_and_deploy_pool::<T>(
+            buyer.clone(),
+            base_asset,
+            asset_count,
+            _10.saturated_into(),
+        );
+
+        let asset = Asset::CategoricalOutcome(market_id, 0u16);
+        let amount_in = _1000.saturated_into();
+        assert_ok!(T::AssetManager::deposit(base_asset, &buyer, amount_in));
+
+        let max_price = _9_10.saturated_into();
+        let orders = (0u128..o as u128).collect::<Vec<_>>();
+        let maker_asset = asset;
+        let maker_amount = _1000.saturated_into();
+        let taker_asset = base_asset;
+        let taker_amount = _1_100.saturated_into();
+        for order_id in &orders {
+            let order_creator: T::AccountId = account("order_creator", *order_id as u32, 0);
+            assert_ok!(T::AssetManager::deposit(maker_asset, &order_creator, maker_amount));
+            T::OrderBook::place_order(
+                &order_creator,
+                market_id,
+                maker_asset,
+                maker_amount,
+                taker_asset,
+                taker_amount,
+            )
+            .unwrap();
```

Comment:
Bit confused by this setup. The maker amount is 1,000, the taker amount is 1/100. This means the price of these orders is a fraction of a penny. Thus, the execution path of `buy` would be:

- Fill all these orders.
- Buy from the AMM until a price of 0.9 is reached.
- Place a limit order.

But the worst-case scenario that I can think of is to spread the orders out, let's say in intervals of a tenth of a cent (so you have one order at the price of 0.1, then one order at the price 0.101, one at 0.102, etc.). This would result in the following execution path:

- Fill one order.
- Buy from the AMM until the price of the next order is reached.
- Repeat `o` times.
- Place a limit order.

Seems considerably worse since you're going from O(1) executions on the AMM to O(n) executions.

Reply:
1dca9663963985bef4cd0c62417f7eacda6c94d1